File: packages/json/llm_json_stream_dart/lib/src/number_property_delegate.dart
import 'property_delegate.dart';

class NumberPropertyDelegate extends PropertyDelegate {
  final StringBuffer _buffer = StringBuffer();

  NumberPropertyDelegate({
    required super.propertyPath,
    required super.parserController,
    super.onComplete,
  });

  @override
  void addCharacter(String character) {
    // Check if this is a delimiter that ends the number
    if (character == "," ||
        character == "}" ||
        character == "]" ||
        character == " " ||
        character == "\n" ||
        character == "\r" ||
        character == "\t") {
      if (_buffer.isNotEmpty && !isDone) {
        isDone = true; // Mark done BEFORE completing to prevent re-entry
        _completeNumber();
        onComplete?.call();
      }
      // Don't consume the character - let the parent handle delimiters
      // The parent will reprocess this character after seeing isDone=true
      return;
    }

    // Valid number characters: digits, minus sign, decimal point, exponent
    if (_isValidNumberCharacter(character)) {
      _buffer.write(character);
    }
  }

  bool _isValidNumberCharacter(String character) {
    return character == '-' ||
        character == '+' ||
        character == '.' ||
        character == 'e' ||
        character == 'E' ||
        (character.codeUnitAt(0) >= 48 && character.codeUnitAt(0) <= 57); // 0-9
  }

  void _completeNumber() {
    if (_buffer.isEmpty) return; // Defensive check

    final numberString = _buffer.toString();
    final number = num.parse(numberString);

    parserController.addPropertyChunk<num>(
      propertyPath: propertyPath,
      chunk: number,
    );

    _buffer.clear(); // Clear buffer to prevent re-use
    // Note: addPropertyChunk already completes the controller, so we don't need to call complete again
  }

  @override
  void onChunkEnd() {
    // Numbers don't emit partial chunks like strings do
  }
}

File: packages/json/llm-json-stream-ts/packages/llm-json-stream/src/classes/property_delegates/number_property_delegate.ts
/**
 * Delegate for parsing JSON number values.
 * Handles integers, floating point, and scientific notation.
 */

import { PropertyDelegate } from "./property_delegate.js";
import { JsonStreamParserController } from "../json_stream_parser.js";

export class NumberPropertyDelegate extends PropertyDelegate {
    private buffer = "";

    constructor(
        propertyPath: string,
        parserController: JsonStreamParserController,
        onComplete?: () => void,
    ) {
        super(propertyPath, parserController, onComplete);
    }

    addCharacter(character: string): void {
        // Check if this is a delimiter that ends the number
        if (
            character === "," ||
            character === "}" ||
            character === "]" ||
            character === " " ||
            character === "\n" ||
            character === "\r" ||
            character === "\t"
        ) {
            if (this.buffer.length > 0 && !this.isDone) {
                this.isDone = true; // Mark done BEFORE completing to prevent re-entry
                this.completeNumber();
                this.onComplete?.();
            }
            // Don't consume the character - let the parent handle delimiters
            // The parent will reprocess this character after seeing isDone=true
            return;
        }

        // Valid number characters: digits, minus sign, decimal point, exponent
        if (this.isValidNumberCharacter(character)) {
            this.buffer += character;
        }
    }

    private isValidNumberCharacter(character: string): boolean {
        return (
            character === "-" ||
            character === "+" ||
            character === "." ||
            character === "e" ||
            character === "E" ||
            (character >= "0" && character <= "9")
        );
    }

    private completeNumber(): void {
        if (this.buffer.length === 0) return; // Defensive check

        const number = parseFloat(this.buffer);

        // Complete the controller directly (number emits once)
        this.parserController.completeProperty(this.propertyPath, number);

        this.buffer = ""; // Clear buffer to prevent re-use
    }

    override onChunkEnd(): void {
        // Numbers don't emit partial chunks like strings do
    }
}
